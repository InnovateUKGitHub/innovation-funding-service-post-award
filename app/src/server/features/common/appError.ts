import { ErrorCode, IAppError } from "@framework/types";
import { Results } from "../../../ui/validation/results";

export class AppError extends Error implements IAppError {
  public results: Results<{}> | null = null;

  constructor(public code: ErrorCode, public message: string, public original?: Error) {
    super();
  }
}

export class NotFoundError extends AppError {
  constructor(
    details?: string,
    readonly original?: Error
  ) {
    super(ErrorCode.REQUEST_ERROR, details || "Not Found", original);
  }
}

export class ForbiddenError extends AppError {
  constructor(
    details?: string,
    readonly original?: Error
  ) {
    super(ErrorCode.FORBIDDEN_ERROR, details || "Forbidden", original);
  }
}

export class FormHandlerError extends AppError {
  constructor(
    public key: string, public store: string, public dto: {}, public result: Results<{}>, public error: IAppError
  ) {
    super(error.code, error.message || "Not Found");
  }
}

export class BadRequestError extends AppError {
  constructor(
    details?: string,
    readonly original?: Error
  ) {
    super(ErrorCode.BAD_REQUEST_ERROR, details || "Invalid Request", original);
  }
}

export class ValidationError extends AppError {
  constructor(
    results: Results<{}>,
    readonly original?: Error
  ) {
    super(ErrorCode.VALIDATION_ERROR, "Validation Error", original);
    this.results = results;
  }
}

export class UnauthenticatedError extends AppError {
  constructor() {
    super(ErrorCode.UNAUTHENTICATED_ERROR, "User not authenticated");
  }
}

export class ConfigurationError extends AppError {
  constructor(message: string) {
    super(ErrorCode.CONFIGURATION_ERROR, message);
  }
}
