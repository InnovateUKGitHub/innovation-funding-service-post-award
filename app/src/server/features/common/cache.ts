import { ICache } from "@framework/types";

export class Cache<T> implements ICache<T> {
  private readonly store: { [key: string]: T } = {};
  private readonly timeouts: { [key: string]: NodeJS.Timer } = {};

  constructor(private readonly minutes: number) {}

  fetch(key: string, get: () => T): T {
    const result = this.store[key];
    if (result) {
      return result;
    }
    return this.set(key, get());
  }

  async fetchAsync(key: string, get: () => Promise<T>): Promise<T> {
    const result = this.store[key];
    if (result) {
      return Promise.resolve(result);
    }
    return this.set(key, await get());
  }

  set(key: string, item: T): T {
    this.clear(key);
    this.store[key] = item;
    if(this.minutes) {
      this.timeouts[key] = setTimeout(() => this.clear(key), this.minutes * 60 * 1000);
    }
    return item;
  }

  clear(key: string) {
    if (this.timeouts[key]) {
      clearTimeout(this.timeouts[key]);
      delete this.timeouts[key];
    }
    delete this.store[key];
  }
}
