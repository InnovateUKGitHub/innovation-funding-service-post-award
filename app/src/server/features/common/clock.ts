import { DateTime } from "luxon";

export const salesforceDateFormat = "yyyy-MM-dd";

export interface IClock {
  now(): Date;
  parse(value: string | null | undefined, format: string): Date | null;
  dateTime(value: Date): DateTime;
  dateTime(value: string, format: string): DateTime;
  parseOptionalSalesforceDate(value: string | null): Date | null;
  formatOptionalSalesforceDate(jsDate?: Date | null): string | null;
  parseRequiredSalesforceDate(value: string): Date;
  parseOptionalSalesforceDateTime(value: string): Date | null;
  parseRequiredSalesforceDateTime(value: string): Date;
}

export class Clock implements IClock {
  now() {
    return new Date();
  }

  parseOptionalSalesforceDate(value: string| null): Date | null {
    if (!value) return null;
    const result = DateTime.fromFormat(value, salesforceDateFormat).setZone("Europe/London");
    if (!result.isValid) {
      throw new Error("Invalid Date: " + value);
    }
    return result.toJSDate();
  }

  formatOptionalSalesforceDate(jsDate?: Date | null): string | null {
    return jsDate ? DateTime.fromJSDate(jsDate).toFormat(salesforceDateFormat) : null;
  }

  parseRequiredSalesforceDate(value: string): Date {
    const result = this.parseOptionalSalesforceDate(value);
    if (!result) {
      throw new Error("Invalid Date: empty");
    }
    return result;
  }

  parseOptionalSalesforceDateTime(value: string|null): Date | null {
    if (!value) return null;
    const result = DateTime.fromISO(value);
    if (!result.isValid) {
      throw new Error("Invalid DateTime: " + value);
    }
    return result.toJSDate();
  }

  parseRequiredSalesforceDateTime(value: string): Date {
    const result = this.parseOptionalSalesforceDateTime(value);
    if (!result) {
      throw new Error("Invalid DateTime: empty");
    }
    return result;
  }

  parse(value: string | null | undefined, format: string): Date | null {
    if (!value) {
      return null;
    }

    return this.dateTime(value, format).toJSDate();
  }

  dateTime(value: Date | string, format?: string): DateTime {
    const localOpts = { locale: "en-GB", zone: "Europe/London" };

    return typeof value === "string"
      ? DateTime.fromFormat(value, format!, localOpts)
      : DateTime.fromJSDate(value, localOpts);
  }
}
