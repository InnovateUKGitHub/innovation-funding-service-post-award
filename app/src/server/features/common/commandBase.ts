import { Authorisation, IContext } from "../../../types";

export abstract class CommandBase<T> {
  protected abstract Run(context: IContext): Promise<T>;

  protected accessControl(auth: Authorisation, context: IContext) {
    return Promise.resolve(true);
  }

  protected LogMessage(): any[] {
    return [this];
  }
}

export abstract class SyncCommandBase<T> {
  protected abstract Run(context: IContext): T;

  protected LogMessage(): any[] {
    return [this];
  }
}
