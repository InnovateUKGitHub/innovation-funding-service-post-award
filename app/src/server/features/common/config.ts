import bytes from "bytes";
import { parseLogLevel } from "@framework/types/logLevel";
import { parseAccEnvironment } from "@framework/types/accEnvironment";
import { IConfig } from "@framework/types/IConfig";
import { AccEnvironment, LogLevel } from "@framework/constants/enums";
import {
  getBooleanEnv as boolEnv,
  getCertificateEnv as certEnv,
  getCommaSepStringArray as csvEnv,
  getEnv as env,
  getFloatEnv as floatEnv,
  getIntegerEnv as intEnv,
  getPipeSepStringArray as psvEnv,
  getStringEnv as strEnv,
} from "@framework/util/envHelpers";

const configuration: Readonly<IConfig> = {
  build: strEnv("BUILD", String(Date.now())),
  salesforceQueryLimit: intEnv("SALESFORCE_QUERY_LIMIT", 195),
  disableCsp: boolEnv("DISABLE_CSP", false),
  cookie: {
    secret: strEnv("COOKIE_KEY"),
    secure: boolEnv("COOKIE_SECURE", true),
  },
  timeouts: {
    costCategories: floatEnv("COST_CAT_TIMEOUT_MINUTES", 720),
    projectRoles: floatEnv("PROJ_ROLES_TIMEOUT_MINUTES", 720),
    recordTypes: floatEnv("RECORD_TYPES_TIMEOUT_MINUTES", 720),
    optionsLookup: floatEnv("OPTIONS_LOOKUP_TIMEOUT_MINUTES", 720),
    token: floatEnv("TOKEN_TIMEOUT_MINUTES", 10),
    cookie: floatEnv("COOKIE_TIMEOUT_MINUTES", 30),
    contentRefreshSeconds: floatEnv("CONTENT_REFRESH_TIMEOUT_SECONDS", 0),
  },
  certificates: {
    salesforce: certEnv("SALESFORCE_PRIVATE_KEY"),
    saml: {
      idp: {
        public: certEnv("SHIBBOLETH_IDP_PUBLIC_KEY", ""),
      },
      spSigning: {
        public: certEnv("SHIBBOLETH_SP_SIGNING_PUBLIC_KEY", ""),
        private: certEnv("SHIBBOLETH_SP_SIGNING_PRIVATE_KEY", ""),
      },
      spDecryption: {
        public: certEnv("SHIBBOLETH_SP_DECRYPTION_PUBLIC_KEY", ""),
        private: certEnv("SHIBBOLETH_SP_DECRYPTION_PRIVATE_KEY", ""),
      },
    },
  },
  basicAuth: {
    credentials: psvEnv("BASIC_AUTH", []),
  },
  features: {
    changePeriodLengthWorkflow: boolEnv("FEATURE_CHANGE_PERIOD_LENGTH", false),
    customContent: boolEnv("FEATURE_CUSTOM_CONTENT", false),
    searchDocsMinThreshold: intEnv("SEARCH_DOCS_MIN_THRESHOLD", 5),
    futureTimeExtensionInYears: intEnv("FUTURE_TIME_EXTENSION_IN_YEARS", 5),
    approveNewSubcontractor: boolEnv("FEATURE_APPROVE_NEW_SUBCONTRACTOR", false),
  },
  logLevel: env("LOG_LEVEL", LogLevel.ERROR, parseLogLevel),
  options: {
    bankCheckAddressScorePass: intEnv("BANK_CHECK_ADDRESS_SCORE_PASS", 6),
    bankCheckValidationRetries: intEnv("BANK_CHECK_VALIDATION_RETRIES", 3),
    bankCheckCompanyNameScorePass: intEnv("BANK_CHECK_COMPANY_NAME_SCORE_PASS", 6),
    permittedTypes: {
      pdfTypes: csvEnv("PERMITTED_DOCUMENT_FILE_TYPES", ["pdf", "xps"]),
      textTypes: csvEnv("PERMITTED_WORD_PROCESSING_FILE_TYPES", ["doc", "docx", "rtf", "txt", "odt"]),
      presentationTypes: csvEnv("PERMITTED_PRESENTATION_FILE_TYPES", ["ppt", "pptx", "odp"]),
      spreadsheetTypes: csvEnv("PERMITTED_SPREADSHEET_FILE_TYPES", ["csv", "xls", "xlsx", "ods"]),
      imageTypes: csvEnv("PERMITTED_IMAGE_FILE_TYPES", ["jpg", "jpeg", "png", "odg"]),
    },
    maxUploadFileCount: intEnv("MAX_UPLOAD_FILE_COUNT", 10),
    maxFileSize: intEnv("MAX_FILE_SIZE_IN_BYTES", bytes("32MB")),
    maxTotalFileSize: intEnv("MAX_TOTAL_FILE_SIZE_IN_BYTES", bytes("32MB")),
    maxFileBasenameLength: intEnv("MAX_FILE_BASENAME_LENGTH", 80),
    standardOverheadRate: floatEnv("STANDARD_OVERHEAD_RATE", 20),
    numberOfProjectsToSearch: intEnv("FEATURE_SEARCH_NUMBER_PROJECTS", 3),
    maxClaimLineItems: intEnv("FEATURE_MAX_CLAIM_LINE_ITEMS", 120),
    nonJsMaxClaimLineItems: intEnv("NON_JS_FEATURE_MAX_CLAIM_LINE_ITEMS", 120),
  },
  salesforceServiceUser: {
    clientId: strEnv("SALESFORCE_CLIENT_ID"),
    connectionUrl: strEnv("SALESFORCE_CONNECTION_URL"),
    serviceUsername: strEnv("SALESFORCE_USERNAME"),
  },
  bankDetailsValidationUser: {
    clientId: strEnv("SALESFORCE_CLIENT_ID"),
    connectionUrl: strEnv("SALESFORCE_CONNECTION_URL"),
    serviceUsername: strEnv("SALESFORCE_BANK_DETAILS_VALIDATION_USERNAME"),
  },
  companiesHouse: {
    endpoint: strEnv("COMPANIES_HOUSE_ENDPOINT", ""),
    accessToken: strEnv("COMPANIES_HOUSE_ACCESS_TOKEN", ""),
  },
  sil: {
    url: strEnv("SIL_HYDRA_URL", ""),
  },
  sso: {
    enabled: boolEnv("USE_SSO", false),
    providerUrl: strEnv("SSO_PROVIDER_URL", ""),
    signoutUrl: strEnv("SSO_SIGNOUT_URL", ""),
  },
  urls: {
    ifsRoot: strEnv("IFS_ROOT", "https://apply-for-innovation-funding.service.gov.uk"),
    ifsApplicationUrl: strEnv(
      "IFS_APPLICATION_URL",
      "https://application-for-innovation-funding.service.gov.uk/management/competition/<<Acc_CompetitionId__c>>/application/<<Acc_IFSApplicationId__c>>",
    ),
    ifsGrantLetterUrl: strEnv(
      "IFS_GRANT_LETTER_URL",
      "https://application-for-innovation-funding.service.gov.uk/management/competition/<<Acc_CompetitionId__c>>/project/<<Acc_IFSApplicationId__c>>",
    ),
  },
  googleTagManagerCode: strEnv("GOOGLE_TAG_MANAGER_CODE", ""),
  s3Account: {
    accessKeyId: strEnv("AWS_S3_ACCESS_KEY_ID", ""),
    secretAccessKey: strEnv("AWS_S3_ACCESS_SECRET", ""),
    contentBucket: strEnv("AWS_S3_CONTENT_BUCKET", ""),
    customContentPath: strEnv("AWS_S3_CONTENT_CUSTOM_PATH", ""),
  },
  accEnvironment: env("ACC_ENVIRONMENT", AccEnvironment.UNKNOWN, parseAccEnvironment),
  newRelic: {
    enabled: boolEnv("NEW_RELIC_ENABLED", false),
    eventsUrl: strEnv("NEW_RELIC_EVENTS_URL", ""),
    apiKey: strEnv("NEW_RELIC_API_KEY", ""),
    appName: strEnv("NEW_RELIC_APP_NAME", ""),
  },
  webserver: {
    port: strEnv("PORT", "8080"),
    url: strEnv("SERVER_URL"),
  },
  developer: {
    writeGraphQL: boolEnv("DEVELOPER_WRITE_GRAPHQL", false),
  },
};

export { configuration };
