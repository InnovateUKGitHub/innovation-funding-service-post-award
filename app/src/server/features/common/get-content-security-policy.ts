// https://developers.google.com/web/fundamentals/security/csp#policy_applies_to_a_wide_variety_of_resources
const cspSourceKeyWords: Readonly<string[]> = ["none", "self", "unsafe-inline", "unsafe-eval"];
const quotedKeywords: Readonly<string[]> = [...cspSourceKeyWords, "nonce", "strict-dynamic"];

type CspPolicySources =
  | "default-src"
  | "base-uri"
  | "object-src"
  | "script-src"
  | "style-src"
  | "img-src"
  | "connect-src"
  | "font-src"
  | "report-uri";

export const policyConfig = (nonce: string): Record<CspPolicySources, (string | typeof cspSourceKeyWords[0])[]> => ({
  "default-src": ["self"],
  "base-uri": ["self"],
  "object-src": ["none"],
  "script-src": [
    "self",
    "unsafe-inline",
    "strict-dynamic",
    `nonce-${nonce}`,
    "https://www.googletagmanager.com",
    "https://tagmanager.google.com",
    "https://www.google-analytics.com",
    "https://ssl.google-analytics.com",
  ],
  "style-src": ["self", "unsafe-inline", "https://tagmanager.google.com", "https://fonts.googleapis.com"],
  "img-src": [
    "self",
    "www.googletagmanager.com",
    "https://www.google-analytics.com",
    "https://ssl.gstatic.com",
    "https://www.gstatic.com",
  ],
  "connect-src": ["self", "https://www.google-analytics.com"],
  "font-src": ["self", "https://fonts.gstatic.com"],
  "report-uri": ["/api/csp/violation-report"],
});

export function getContentSecurityPolicy(nonceValue: string): string {
  const policyKeyDelimiter = "; ";
  const cspConfig = policyConfig(nonceValue);

  const configkeys = Object.keys(cspConfig) as CspPolicySources[];

  const getCspValue = (cspValue: string) => {
    const isNonce = cspValue.slice(0, 6) === "nonce-";
    const isQuoteValid = quotedKeywords.includes(cspValue);
    return isNonce || isQuoteValid ? `'${cspValue}'` : cspValue;
  };

  const cspPolicy = configkeys.reduce((partialPolicyString, policyKey) => {
    const policyOptions = cspConfig[policyKey];
    const policyValue = policyOptions.map(getCspValue).join(" ");
    const nextCspPolicy = `${policyKey} ${policyValue}` + policyKeyDelimiter;

    return partialPolicyString.concat(nextCspPolicy);
  }, "");

  return cspPolicy.trim();
}
