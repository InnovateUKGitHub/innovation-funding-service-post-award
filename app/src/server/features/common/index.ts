export * from "./appError";
export * from "./cache";
export * from "./commandBase";
export * from "./config";
export * from "./context";
export * from "./queryBase";
export * from "./timer";
export * from "./get-content-security-policy";
