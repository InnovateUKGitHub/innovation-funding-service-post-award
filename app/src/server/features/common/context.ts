import i18next from "i18next";
import { CustomContentStore } from "@server/resources/customContentStore";
import { CostCategoryDto } from "@framework/dtos/costCategoryDto";
import { BankCheckService } from "@server/resources/bankCheckService";
import { CompaniesHouseBase } from "@server/resources/companiesHouse";
import { GetAllProjectRolesForUser, IRoleInfo } from "../projects/getAllProjectRolesForUser";
import { GetRecordTypeQuery } from "../general/getRecordTypeQuery";
import { AppError, BadRequestError, ForbiddenError, NotFoundError, ValidationError } from "./appError";
import { ILogger, Logger } from "@shared/developmentLogger";
import { ErrorCode } from "@framework/constants/enums";
import { Authorisation } from "@framework/types/authorisation";
import {
  ICaches,
  IContext,
  IRepositories,
  IResources,
  IInternationalisation,
  IAsyncRunnable,
  ISyncRunnable,
} from "@framework/types/IContext";
import { ISessionUser } from "@framework/types/IUser";
import { IClock, Clock } from "@framework/util/clock";
import { AccountsRepository } from "@server/repositories/accountsRepository";
import { ClaimDetailsRepository } from "@server/repositories/claimDetailsRepository";
import { ClaimLineItemRepository } from "@server/repositories/claimLineItemRepository";
import { ClaimRepository } from "@server/repositories/claimsRepository";
import { ClaimStatusChangeRepository } from "@server/repositories/claimStatusChangeRepository";
import { ClaimTotalCostCategoryRepository } from "@server/repositories/claimTotalCostCategoryRepository";
import { CompaniesHouse } from "@server/repositories/companiesRepository";
import { CostCategoryRepository } from "@server/repositories/costCategoriesRepository";
import { DocumentsRepository } from "@server/repositories/documentsRepository";
import {
  SalesforceTokenError,
  SalesforceInvalidFilterError,
  FileTypeNotAllowedError,
} from "@server/repositories/errors";
import { FinancialLoanVirementRepository } from "@server/repositories/financialLoanVirementRepository";
import { FinancialVirementRepository } from "@server/repositories/financialVirementRepository";
import { LoanRepository } from "@server/repositories/loanRepository";
import { MonitoringReportHeaderRepository } from "@server/repositories/monitoringReportHeaderRepository";
import { MonitoringReportQuestionsRepository } from "@server/repositories/monitoringReportQuestionsRepository";
import { MonitoringReportResponseRepository } from "@server/repositories/monitoringReportResponseRepository";
import { MonitoringReportStatusChangeRepository } from "@server/repositories/monitoringReportStatusChangeRepository";
import { PartnerRepository } from "@server/repositories/partnersRepository";
import { PcrSpendProfileRepository } from "@server/repositories/pcrSpendProfileRepository";
import { PermissionGroupRepository } from "@server/repositories/permissionGroupsRepository";
import { ProfileDetailsRepository } from "@server/repositories/profileDetailsRepository";
import { ProfileTotalPeriodRepository } from "@server/repositories/profilePeriodTotalRepository";
import { ProfileTotalCostCategoryRepository } from "@server/repositories/profileTotalCostCategoryRepository";
import { ProjectChangeRequestRepository } from "@server/repositories/projectChangeRequestRepository";
import { ProjectChangeRequestStatusChangeRepository } from "@server/repositories/projectChangeRequestStatusChangeRepository";
import { ProjectContactsRepository } from "@server/repositories/projectContactsRepository";
import { ProjectRepository } from "@server/repositories/projectsRepository";
import { RecordTypeRepository } from "@server/repositories/recordTypeRepository";
import { CommandBase, NonAuthorisedCommandBase, SyncCommandBase } from "./commandBase";
import { configuration } from "./config";
import { IConfig } from "@framework/types/IConfig";
import { QueryBase, SyncQueryBase } from "./queryBase";
import { Timer } from "./timer";
import { Option } from "@framework/dtos/option";
import { Cache } from "./cache";
import { PermissionGroup } from "@framework/entities/permissionGroup";
import { RecordType } from "@framework/entities/recordType";
import { ISalesforceConnectionDetails, salesforceConnectionWithToken } from "@server/repositories/salesforceConnection";

// obviously needs to be singleton
const cachesImplementation: ICaches = {
  costCategories: new Cache<CostCategoryDto[]>(configuration.timeouts.costCategories),
  optionsLookup: new Cache<Option<unknown>[]>(configuration.timeouts.optionsLookup),
  projectRoles: new Cache<{ [key: string]: IRoleInfo }>(configuration.timeouts.projectRoles),
  permissionGroups: new Cache<PermissionGroup[]>(0 /* permanent cache */),
  recordTypes: new Cache<RecordType[]>(configuration.timeouts.recordTypes),
  contentStoreLastUpdated: null,
};

export const constructErrorResponse = (error: unknown): AppError => {
  if (
    error instanceof ValidationError ||
    error instanceof ForbiddenError ||
    error instanceof BadRequestError ||
    error instanceof NotFoundError
  ) {
    return error;
  }

  if (error instanceof SalesforceTokenError) {
    return new AppError(ErrorCode.SECURITY_ERROR, error.message, error);
  }

  if (error instanceof SalesforceInvalidFilterError) {
    return new NotFoundError(undefined, error);
  }

  if (error instanceof FileTypeNotAllowedError) {
    return new AppError(ErrorCode.BAD_REQUEST_ERROR, error.message, error);
  }

  if (error instanceof Error) {
    // TODO capture stack trace for logs
    return new AppError(ErrorCode.UNKNOWN_ERROR, error.message, error);
  }

  if (typeof error === "string") {
    return new AppError(ErrorCode.UNKNOWN_ERROR, error);
  }

  // We don't really know what type the error is, so force it into a string.
  return new AppError(ErrorCode.UNKNOWN_ERROR, JSON.stringify(error));
};

export class Context implements IContext {
  constructor(public readonly user: ISessionUser) {
    this.config = configuration;

    const salesforceConfig = {
      clientId: this.config.salesforceServiceUser.clientId,
      connectionUrl: this.config.salesforceServiceUser.connectionUrl,
      serviceUsername: this.config.salesforceServiceUser.serviceUsername,
    };

    this.salesforceConnectionDetails = {
      ...salesforceConfig,
      currentUsername: this.user?.email,
    };

    this.logger = new Logger("Context", {
      prefixLines: [user && user.email],
    });

    this.caches = cachesImplementation;

    // use fat arrow so this is bound - extracted to shorten line length
    const connectionCallback = () => this.getSalesforceConnection();
    const asSystemUserConnectionCallback = () => this.asSystemUser().getSalesforceConnection();
    const recordTypeCallback = (objectName: string, recordType: string) => this.getRecordTypeId(objectName, recordType);

    this.repositories = {
      accounts: new AccountsRepository(connectionCallback, this.logger),
      claims: new ClaimRepository(connectionCallback, this.logger),
      claimDetails: new ClaimDetailsRepository(recordTypeCallback, connectionCallback, this.logger),
      claimStatusChanges: new ClaimStatusChangeRepository(connectionCallback, this.logger),
      companies: new CompaniesHouse(),
      claimTotalCostCategory: new ClaimTotalCostCategoryRepository(connectionCallback, this.logger),
      claimLineItems: new ClaimLineItemRepository(recordTypeCallback, connectionCallback, this.logger),
      costCategories: new CostCategoryRepository(connectionCallback, this.logger),
      documents: new DocumentsRepository(connectionCallback, asSystemUserConnectionCallback, this.logger),
      financialVirements: new FinancialVirementRepository(recordTypeCallback, connectionCallback, this.logger),
      financialLoanVirements: new FinancialLoanVirementRepository(recordTypeCallback, connectionCallback, this.logger),
      pcrSpendProfile: new PcrSpendProfileRepository(recordTypeCallback, connectionCallback, this.logger),
      monitoringReportResponse: new MonitoringReportResponseRepository(
        recordTypeCallback,
        connectionCallback,
        this.logger,
      ),
      monitoringReportHeader: new MonitoringReportHeaderRepository(recordTypeCallback, connectionCallback, this.logger),
      monitoringReportQuestions: new MonitoringReportQuestionsRepository(connectionCallback, this.logger),
      loans: new LoanRepository(connectionCallback, this.logger),
      monitoringReportStatusChange: new MonitoringReportStatusChangeRepository(connectionCallback, this.logger),
      projectChangeRequests: new ProjectChangeRequestRepository(recordTypeCallback, connectionCallback, this.logger),
      profileDetails: new ProfileDetailsRepository(connectionCallback, this.logger),
      profileTotalPeriod: new ProfileTotalPeriodRepository(connectionCallback, this.logger),
      profileTotalCostCategory: new ProfileTotalCostCategoryRepository(connectionCallback, this.logger),
      projects: new ProjectRepository(connectionCallback, this.logger),
      partners: new PartnerRepository(connectionCallback, this.logger),
      projectChangeRequestStatusChange: new ProjectChangeRequestStatusChangeRepository(connectionCallback, this.logger),
      projectContacts: new ProjectContactsRepository(connectionCallback, this.logger),
      permissionGroups: new PermissionGroupRepository(connectionCallback, this.logger),
      recordTypes: new RecordTypeRepository(connectionCallback, this.logger),
    };

    this.resources = {
      bankCheckService: new BankCheckService(),
      companiesHouse: new CompaniesHouseBase(),
      customContent: new CustomContentStore(),
    };
  }

  public readonly repositories: IRepositories;
  public readonly logger: ILogger;
  public readonly config: IConfig;
  public readonly clock: IClock = new Clock();
  public readonly caches: ICaches;
  public readonly resources: IResources;

  public readonly internationalisation: IInternationalisation = {
    addResourceBundle: (content, namespace) => i18next.addResourceBundle("en-GB", namespace, content, true, true),
  };

  private readonly salesforceConnectionDetails: ISalesforceConnectionDetails;

  public getSalesforceConnection() {
    return salesforceConnectionWithToken(this.salesforceConnectionDetails);
  }

  public startTimer(message: string) {
    return new Timer(this.logger, message);
  }

  private authorisation: Authorisation | null = null;

  private getAuthorisation() {
    if (this.authorisation) {
      return Promise.resolve(this.authorisation);
    }
    return new GetAllProjectRolesForUser().run(this).then(x => {
      this.authorisation = x;
      return x;
    });
  }

  private async runAsync<TResult>(runnable: IAsyncRunnable<TResult>): Promise<TResult> {
    const timer = this.startTimer(runnable.constructor.name);
    try {
      if (runnable.accessControl) {
        const authorisation = await this.getAuthorisation();
        if (!(await runnable.accessControl(authorisation, this))) throw new ForbiddenError();
      }
      // await the run because of the finally
      return await runnable.run(this);
    } catch (e: unknown) {
      this.logger.warn("Failed query", runnable.logMessage(), e);
      if (e instanceof ValidationError) {
        this.logger.debug("Validation Error", e.results && e.results.log());
      }
      if (runnable.handleRepositoryError) runnable.handleRepositoryError(this, e);
      throw constructErrorResponse(e);
    } finally {
      timer.finish();
    }
  }

  private runSync<TResult>(runnable: ISyncRunnable<TResult>): TResult {
    const timer = this.startTimer(runnable.constructor.name);
    try {
      return runnable.run(this);
    } catch (e: unknown) {
      this.logger.warn("Failed query", runnable.logMessage(), e);
      throw constructErrorResponse(e);
    } finally {
      timer.finish();
    }
  }

  public runQuery<TResult>(query: QueryBase<TResult>): Promise<TResult> {
    const runnable = query as unknown as IAsyncRunnable<TResult>;
    this.logger.info("Running async query", runnable.logMessage());
    return this.runAsync(runnable);
  }

  public runSyncQuery<TResult>(query: SyncQueryBase<TResult>): TResult {
    const runnable = query as unknown as ISyncRunnable<TResult>;
    this.logger.info("Running sync query", runnable.logMessage());
    return this.runSync(runnable);
  }

  public runCommand<TResult>(command: CommandBase<TResult> | NonAuthorisedCommandBase<TResult>): Promise<TResult> {
    const runnable = command as unknown as IAsyncRunnable<TResult>;
    this.logger.info("Running async command", ...runnable.logMessage());
    return this.runAsync(runnable);
  }

  public runSyncCommand<TResult>(command: SyncCommandBase<TResult>): TResult {
    const runnable = command as unknown as ISyncRunnable<TResult>;
    this.logger.info("Running sync command", runnable.logMessage());
    return this.runSync(runnable);
  }

  /**
   * Elevate a user context to a different user context.
   *
   * @param user The Salesforce user e-mail address
   * @returns An elevated IContext as the passed in user
   * @author Leondro Lio <leondro.lio@iuk.ukri.org>
   */
  private elevateUserAs(user: string): IContext {
    if (this.user.email !== user) {
      return new Context({ email: user });
    }
    return this;
  }

  /**
   * Elevate a user context to the system user context
   *
   * @returns An elevated IContext as the system user
   */
  public asSystemUser(): IContext {
    const serviceUser = this.config.salesforceServiceUser.serviceUsername;
    this.logger.info(`Escalating from ${this.user.email} to system user ${serviceUser}`);
    return this.elevateUserAs(serviceUser);
  }

  /**
   * Elevate a user context to the bank details validation user context.
   *
   * @returns An elevated IContext as the bank details validation user
   */
  public asBankDetailsValidationUser(): IContext {
    const serviceUser = this.config.bankDetailsValidationUser.serviceUsername;
    this.logger.info(`Escalating from ${this.user.email} to banking user ${serviceUser}`);
    return this.elevateUserAs(serviceUser);
  }

  // helper function for repositories that need record type ids
  private async getRecordTypeId(objectName: string, recordType: string): Promise<string> {
    const query = new GetRecordTypeQuery(objectName, recordType);
    return this.runQuery(query).then(x => x.id);
  }
}
