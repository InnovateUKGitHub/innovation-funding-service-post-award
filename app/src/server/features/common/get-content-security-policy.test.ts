import { getContentSecurityPolicy, policyConfig } from "@server/features/common/get-content-security-policy";

const stubNonce = "STUB_UNIQUE_NONCE_VALUE";

describe("policyConfig()", () => {
  it("returns with config", () => {
    const config = policyConfig(stubNonce);

    expect(config).toMatchSnapshot();
  });
});

describe("getContentSecurityPolicy()", () => {
  describe("@returns", () => {
    test("with expected nonce", () => {
      const stubUniqueNonce = "chicken";
      const config = getContentSecurityPolicy(stubUniqueNonce);

      const expectedNonce = `nonce-${stubUniqueNonce}`;

      expect(config).toContain(expectedNonce);
    });

    test("with correct formatting", () => {
      const config = getContentSecurityPolicy(stubNonce);

      expect(config).toMatchSnapshot();
    });

    test("with report-uri as last policy", () => {
      const config = getContentSecurityPolicy(stubNonce);

      // Note: Separate csp policy using delimiter
      const cspPolicies = config.split("; ");
      const lastCspPolicy = cspPolicies[cspPolicies.length - 1];

      expect(lastCspPolicy).toContain("report-uri");
    });
  });
});
