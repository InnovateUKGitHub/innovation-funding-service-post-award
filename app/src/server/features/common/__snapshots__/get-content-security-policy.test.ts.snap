// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getContentSecurityPolicy() @returns with correct formatting 1`] = `"default-src 'self'; base-uri 'self'; object-src 'none'; script-src 'self' 'unsafe-eval' 'strict-dynamic' 'nonce-STUB_UNIQUE_NONCE_VALUE' https://www.googletagmanager.com https://tagmanager.google.com https://*.google-analytics.com https://tagassistant.google.com/; style-src 'self' 'unsafe-inline' https://tagmanager.google.com https://fonts.googleapis.com https://www.googletagmanager.com; img-src 'self' data: https://www.googletagmanager.com https://www.google-analytics.com https://ssl.gstatic.com https://www.gstatic.com https://fonts.gstatic.com; frame-src https://www.googletagmanager.com; connect-src 'self' https://*.google-analytics.com; font-src 'self' https://assets.publishing.service.gov.uk/frontend/ https://fonts.gstatic.com; report-uri /api/csp/violation-report;"`;

exports[`policyConfig() returns with config 1`] = `
{
  "base-uri": [
    "self",
  ],
  "connect-src": [
    "self",
    "https://*.google-analytics.com",
  ],
  "default-src": [
    "self",
  ],
  "font-src": [
    "self",
    "https://assets.publishing.service.gov.uk/frontend/",
    "https://fonts.gstatic.com",
  ],
  "frame-src": [
    "https://www.googletagmanager.com",
  ],
  "img-src": [
    "self",
    "data:",
    "https://www.googletagmanager.com",
    "https://www.google-analytics.com",
    "https://ssl.gstatic.com",
    "https://www.gstatic.com",
    "https://fonts.gstatic.com",
  ],
  "object-src": [
    "none",
  ],
  "report-uri": [
    "/api/csp/violation-report",
  ],
  "script-src": [
    "self",
    "unsafe-eval",
    "strict-dynamic",
    "nonce-STUB_UNIQUE_NONCE_VALUE",
    "https://www.googletagmanager.com",
    "https://tagmanager.google.com",
    "https://*.google-analytics.com",
    "https://tagassistant.google.com/",
  ],
  "style-src": [
    "self",
    "unsafe-inline",
    "https://tagmanager.google.com",
    "https://fonts.googleapis.com",
    "https://www.googletagmanager.com",
  ],
}
`;
